### ISSUE TITLE
MemoryRecord struct is not public visible. This blocks extending the storage system

### ISSUE THREAD

**Issue by JohnMasen** [microsoft/semantic-kernel#202]
I have a large number of text files already embedded, embedding vectors are stored in a single file (embedding.jsonz). I would like to add these files to the memory system. However It turns out I could not reuse SemanticTextMemory class with my own IMemoryStore<float> implementation. 
The reason is  SemanticTextMemory expects the result from storage is MemoryRecord, which is an internal type. my storage system could not create this type inside my code.
suggestions:
1. Make MemoryRecord class public.
2. Re-design the memory system. Current memory system is heavly over-cooked.  Too many nested generic interface. As storage system heavyly rely on data serialize/deserialize and there's 5 key elements total (collection,key,content,content-type,embedding). A simple and lightweight data structure would be much better.
Thanks

**Comment by dluc** [comment]
@JohnMasen thanks for raising the issue, I can see that navigating that part of the SDK can be challenging. I think it would be useful providing an example to cover this scenario. @awharrison-28 could we reproduce this scenario and verify that one can actually bring their own memory data, either embeddings, raw data or both?

**Comment by JohnMasen** [comment]
suggested storage interface
``` C#
public struct MemoryRecord
{
    public string? Collection;//null=default collection
    public string Key;
    public string Value;
    public string Value_ContentType; //default is text/plain (string), could be others, we can leave this for further use
    public readonlyspan<float>?  Embedding; //this coul be empty if retrieve from storage
}

Public Interface IMemoryStorage
{
    void InsertOrUpdate(MemoryRecord value);
    MemoryRecord Retrieve(string collection, string key);
    void Remove(string collection,string key);
    void Purge(string collection);
    void InsertOrUpdateBatch(iEnumertable<MemroyRecord> itemList); 
   IEnumertable<MemoryRecord> Query(readonlyspan<float> queryEmbedding, float relevance=0.75f, int top=1);
}
```

**Comment by awharrison-28** [comment]
Hi @JohnMasen, thank you for your feedback. I have hit a similar challenges and have been working on the story for passing vector metadata between storage schemas.

- https://github.com/microsoft/semantic-kernel/pull/210 : reference for how we currently expect IMemoryStore to be implemented. Note that the PR contains a change to IEmbeddingWithMetadata expecting a string representation of the metadata (dotnet/src/SemanticKernel/Memory). 
- your suggestion of `public string Value_ContentType` is also high priority. I agree that it is very important for users to provide custom metadata to their storage not necessarily enforced by the kernel. This will make your suggestion of `IMemoryStore : IDataStore<IEmbeddingWithMetadata>` -> `IMemoryStore : IDataStore<MemoryRecord>` much easier. 

We also understand that the nested generics are noisy. It's certainly something we want to address in a cleaner fashion where possible. 

TLDR; Lots of improvements to Memory are in work. I'll keep this thread updated with relevant PRs.

**Comment by JohnMasen** [comment]
Glad to see such a quick response. Expecting the new version :)

**Comment by awharrison-28** [comment]
@JohnMasen A number of the points you called out are addressed in https://github.com/microsoft/semantic-kernel/pull/308. Note that MemoryRecord is already public as of https://github.com/microsoft/semantic-kernel/pull/250. Adding a `value string` to MemoryRecord is coming soon among some other QoL changes to SemanticTextMemory.

**Comment by JohnMasen** [comment]
That's great!  very appreacted for your guys hard work!

**Comment by JohnMasen** [comment]
I found the latest nuget package (0.10.72.1-preview) is not using the lastest change. Will these changes be included in the next version?

**Comment by awharrison-28** [comment]
@JohnMasen PR is out that adds an optional valueString field to MemoryRecord. IMemoryStore get/search methods now optionally return the embedding to the user too (defaults to false). 

Once those changes propagate to the latest nuget package, can we consider this issue closed?

**Comment by JohnMasen** [comment]
@awharrison-28 By upgrading to the lastest package, I confirm the fix is done, please feel free to close the issue.
Great job, guys!

**Comment by awharrison-28** [comment]
@JohnMasen following up here, the next package release adds a field to MemoryRecordMetadata for custom metadata strings and Get/Search memories calls can optionally return the actual embedding (defaulting to not).
